src/core/midi/MidiApple.cpp

```diff
@@ -0,0 +1,590 @@
+    ItemCount deviceCount = MIDIGetNumberOfDevices();
+    
+    // Iterate through all MIDI devices
+    for (ItemCount i = 0 ; i < deviceCount ; ++i) {
+        
+        // Grab a reference to current device
+        MIDIDeviceRef device = MIDIGetDevice(i);
+        char * deviceName = getName(device);
+        QString qsDeviceName = QString::fromUtf8((char*)(deviceName));
+        qDebug("Device name:%s",deviceName);
+        
+        // Is this device online? (Currently connected?)
+        SInt32 isOffline = 0;
+        MIDIObjectGetIntegerProperty(device, kMIDIPropertyOffline, &isOffline);
+        qDebug(" is online: %s", (isOffline ? "No" : "Yes"));
+        // How many entities do we have?
+        ItemCount entityCount = MIDIDeviceGetNumberOfEntities(device);
+        
+        // Iterate through this device's entities
+        for (ItemCount j = 0 ; j < entityCount ; ++j) {
+            
+            // Grab a reference to an entity
+            MIDIEntityRef entity = MIDIDeviceGetEntity(device, j);
+            qDebug("  Entity: %s", getName(entity));
+            
+            // Iterate through this device's source endpoints (MIDI In)
+            ItemCount sourceCount = MIDIEntityGetNumberOfSources(entity);
+            for (ItemCount k = 0 ; k < sourceCount ; ++k) {
+                
+                // Grab a reference to a source endpoint
+                MIDIEndpointRef source = MIDIEntityGetSource(entity, k);
+                char * name = getName(source);
+                qDebug("    Source: '%s'", name);
+                QString sourceName = qsDeviceName + ":" + QString::fromUtf8((char*)(name));
+                qDebug("    Source name: '%s'", sourceName.toLatin1().constData() );
+                m_inputDevices.insert(sourceName, source);
+                openMidiReference(source,sourceName,true);
+            }
+            
+            // Iterate through this device's destination endpoints (MIDI Out)
+            ItemCount destCount = MIDIEntityGetNumberOfDestinations(entity);
+            for (ItemCount k = 0 ; k < destCount ; ++k) {
+                
+                // Grab a reference to a destination endpoint
+                MIDIEndpointRef dest = MIDIEntityGetDestination(entity, k);
+                char * name = getName(dest);
+                qDebug("    Destination: '%s'", name);
+                QString destinationName = qsDeviceName + ":" + QString::fromUtf8((char*)(name));
+                qDebug("    Destination name: '%s'", destinationName.toLatin1().constData() );
+                m_outputDevices.insert(destinationName, dest);
+                openMidiReference(dest,destinationName,false);
+            }
+        }
+        qDebug("------");
+    }
+    
+    printQStringKeys("m_inputDevices:",m_inputDevices);
+    printQStringKeys("m_outputDevices:",m_outputDevices);
+}
+
+void MidiApple::printQStringKeys(char const * mapName, QMap<QString,MIDIEndpointRef> &m_input){
```

Comment:
Usually m_ prefix is reserved for member variables, not function parameters, so this could lead to confusion if kept as-is. Additionally, the argument should probably  passed by const reference and the function itself can be const.